### PR TITLE
 [clang-tidy] Add user-defined functions to `bugprone-unsafe-functions` check

### PR DESCRIPTION
**Pull request by Discookie**
Adds the check option `bugprone-unsafe-functions.CustomFunctions` to be able to match user-defined functions as part of the checker.
Adds the option `bugprone-unsafe-functions.ReportDefaultFunctions` to disable reporting the default set of functions as well.

The functions names are matched using the same mechanism as the `matchesAnyListedName` tidy matcher, documented in `unsafe-functions.rst`.

### CODE REVIEW

File: clang-tools-extra/clang-tidy/bugprone/UnsafeFunctionsCheck.cpp [hunk 1]
```diff
@@ -7,6 +7,8 @@
 //===----------------------------------------------------------------------===//
 
 #include "UnsafeFunctionsCheck.h"
+#include "../utils/Matchers.h"
```
Comment (whisperity) [inline clang-tools-extra/clang-tidy/bugprone/UnsafeFunctionsCheck.cpp:10]: (This include is superfluous, ClangTidyCheck automatically pulls the related header in. Yes, I'm always burnt by clangd with this as well. :wink:)

File: clang-tools-extra/clang-tidy/utils/Matchers.h [hunk 2]
```diff
@@ -136,6 +128,15 @@ class MatchesAnyListedNameMatcher
...
 
+  bool matches(
+      const NamedDecl &Node, ast_matchers::internal::ASTMatchFinder *Finder,
+      ast_matchers::internal::BoundNodesTreeBuilder *Builder) const override {
+    return llvm::any_of(NameMatchers, [&Node](const NameMatcher &NM) {
+      return NM.match(Node);
+    });
+  }
+
+private:
```
Comment (whisperity) [inline clang-tools-extra/clang-tidy/utils/Matchers.h:139]: Why was this change needed? The other changes in the code do not indicate this function getting called anywhere.

**Comment by whisperity** [comment]
High-level design question: Why is `CustomAnnexKFunctions` a thing?

File: clang-tools-extra/clang-tidy/utils/Matchers.h (hunk 2 above)
Comment (Discookie) [reply clang-tools-extra/clang-tidy/utils/Matchers.h:139]: The diff here highlights the wrong part of the change:
https://github.com/llvm/llvm-project/pull/106350/commits/47c1a5694fed5769bb6bbc437e31276efe41d520

I made the contained NameMatcher class public, but the diff considered the other function to be moved.

**Comment by Discookie** [comment]
`CustomAnnexKFunctions` exists because the checker has two different internal matching modes for whether AnnexK is enabled or not. This both depends on `__STDC_LIB_EXT1__` (defined by system) `__STDC_WANT_LIB_EXT1__` (defined by the user), so it's reasonable to expect that a project could have differing availability of AnnexK.

I could rename CustomNormalFunctions to CustomFunctions or something to make it clearer that it's always available.

File: clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst [hunk 3]
```diff
@@ -74,6 +77,40 @@ Both macros have to be defined to suggest replacement functions from *Annex K.*
...
+     functionRegex1, replacement1[, reason1]; 
+     functionRegex2, replacement2[, reason2];
+     ...
+   "
+
+The functions are matched using POSIX extended regular expressions.
+*(Note: The regular expressions do not support negative* ``(?!)`` *matches)*
+
+The `reason` is optional and is used to provide additional information about the
+reasoning behind the replacement. The default reason is ``is marked as unsafe``.
```
Comment (EugeneZelenko) [inline clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst:99]:
```suggestion
reasoning behind the replacement. The default reason is `is marked as unsafe`.
```

Option values should be enclosed in single back-ticks. Same below.

**Review by EugeneZelenko — COMMENTED** [review]
Please mention changes in Release Notes.

File: clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst [hunk 4]
```diff
@@ -86,6 +123,22 @@ Options
...
+
+    A comma-separated list of regular expressions, their replacements, and an
+    optional reason. For more information, see `Custom functions
+    <unsafe-functions.html#custom-functions>`_.
+
+.. option:: CustomAnnexKFunctions
+
+    A comma-separated list of regular expressions, their replacements, and an
+    optional reason. For more information, see `Custom functions
+    <unsafe-functions.html#custom-functions>`_.
```
Comment (5chmidti) [inline clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst:141]: What do you think about:

> A semicolon-separated list of specifications, each containing a regular expression, a replacement, and an optional reason, separated by a comma.

? I think there is still room for improvement, but I'm worried about the not mentioned semicolon/structure (even though you link to the documentation).

**Review by 5chmidti — COMMENTED** [review]
Just some small things

File: clang-tools-extra/clang-tidy/bugprone/UnsafeFunctionsCheck.cpp [hunk 5]
```diff
@@ -127,57 +139,157 @@ static bool isAnnexKAvailable(std::optional<bool> &CacheVar, Preprocessor *PP,
...
+                                 "expected the name of an unsafe function")
+          << OptionName;
+    }
+
+    if (Replacement.trim().empty()) {
+      Context->configurationDiag(
+          "invalid configuration value '%0' for option '%1'; "
+          "expected a replacement function name")
+          << Name.trim() << OptionName;
+    }
```
[the remark this answers is not in the document]
Comment (Discookie) [reply clang-tools-extra/clang-tidy/bugprone/UnsafeFunctionsCheck.cpp]: For some reason I thought this call was treated as a fatal error, fixed.

File: clang-tools-extra/clang-tidy/bugprone/UnsafeFunctionsCheck.cpp [hunk 6]
```diff
@@ -186,16 +298,62 @@ void UnsafeFunctionsCheck::check(const MatchFinder::MatchResult &Result) {
...
+
+    for (const auto &Entry : CustomNormalFunctions) {
+
+      if (Entry.Pattern.match(*FuncDecl)) {
+        ShowCheckedFunctionWarning(Entry);
+        return;
+      }
+    }
+
+    assert(false && "No custom function was matched.");
```
[the remark this answers is not in the document]
Comment (Discookie) [reply clang-tools-extra/clang-tidy/bugprone/UnsafeFunctionsCheck.cpp]: Only this one, the others have proper invariants attached.

File: clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst (hunk 4 above)
Comment (Discookie) [reply clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst]: > A semicolon-separated list of custom functions to be matched. A matched
    function contains a regular expression, the name of the replacement
    function, and an optional reason, separated by comma.

Should be good enough for these purposes.

File: clang-tools-extra/clang-tidy/bugprone/UnsafeFunctionsCheck.cpp (hunk 1 above)
Comment (Discookie) [reply clang-tools-extra/clang-tidy/bugprone/UnsafeFunctionsCheck.cpp]: Not in this case! utils/Matchers.h is not included in ClangTidyCheck (but it is included in UnsafeFunctionsCheck.h, so... whoops)

File: clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst [hunk 7]
```diff
@@ -19,6 +19,9 @@ The check implements the following rules from the CERT C Coding Standard:
 Unsafe functions
 ----------------
 
+The following functions are reported if `ReportDefaultFunctions
+<unsafe-functions.html#cmdoption-arg-ReportDefaultFunctions>`_ is enabled.
+
```
[the remark this answers is not in the document]
Comment (Discookie) [reply clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst]: I hope I did it correctly, I'm not that familiar with RST syntax.

[the remark this answers is not in the document]
Comment (whisperity) [reply clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst]: Referring to the `.html` directly is definitely not good because rst can be lowered to different markup syntaxes.

You need to create (or check if `opt::` automatically does?) a referable entity, and then `:ref:` can be used to make the link clickable. So you will use the ID or "label" you allocated in the document instead of relying on the automatically generated, HTML-specific anchor.

**Comment by Discookie** [comment]
Internal was the wrong word there, the distinction is only about whether AnnexK is available or not, in the context of the checked code. We certainly could eliminate `CustomAnnexKFunctions`, but that would mean that it's the user's responsibility to check whether AnnexK is available, outside of the code being checked.
Ultimately I don't think it would get too much use, if for nothing other than being able to reproduce the default-checked functions with just the `Custom__Functions`.

File: clang-tools-extra/clang-tidy/bugprone/UnsafeFunctionsCheck.cpp [hunk 8]
```diff
@@ -127,57 +138,159 @@ static bool isAnnexKAvailable(std::optional<bool> &CacheVar, Preprocessor *PP,
   return CacheVar.value();
 }
 
+static std::vector<UnsafeFunctionsCheck::CheckedFunction>
+parseCheckedFunctions(StringRef Option, StringRef OptionName,
+                      ClangTidyContext *Context) {
+  std::vector<StringRef> Functions = utils::options::parseStringList(Option);
```
Comment (EugeneZelenko) [inline clang-tools-extra/clang-tidy/bugprone/UnsafeFunctionsCheck.cpp:144]:
```suggestion
  const std::vector<StringRef> Functions = utils::options::parseStringList(Option);
```

**Comment by Discookie** [comment]
Removed the normal/AnnexK distinction as I think it would get little use as it's a rare edge case, and it made everything more confusing.

**Review by Szelethus — COMMENTED** [review]
Alright, I think this patch is good from a design standpoint! I'll get into the nitty-gritty in a bit.

File: clang-tools-extra/clang-tidy/bugprone/UnsafeFunctionsCheck.cpp [hunk 9]
```diff
@@ -127,57 +134,137 @@ static bool isAnnexKAvailable(std::optional<bool> &CacheVar, Preprocessor *PP,
...
+parseCheckedFunctions(StringRef Option, ClangTidyContext *Context) {
+  const std::vector<StringRef> Functions =
+      utils::options::parseStringList(Option);
+  std::vector<UnsafeFunctionsCheck::CheckedFunction> Result;
+  Result.reserve(Functions.size());
+
+  for (StringRef Function : Functions) {
+    if (Function.empty()) {
+      continue;
+    }
```
Comment (5chmidti) [inline clang-tools-extra/clang-tidy/bugprone/UnsafeFunctionsCheck.cpp:147]: nit: single statement with braces

File: clang-tools-extra/clang-tidy/bugprone/UnsafeFunctionsCheck.cpp [hunk 10]
```diff
@@ -127,57 +134,137 @@ static bool isAnnexKAvailable(std::optional<bool> &CacheVar, Preprocessor *PP,
...
+
+    auto [Name, Rest] = Function.split(',');
+    auto [Replacement, Reason] = Rest.split(',');
```
Comment (5chmidti) [inline clang-tools-extra/clang-tidy/bugprone/UnsafeFunctionsCheck.cpp:150]: nit: Please declare both of these `const`

**Review by 5chmidti — APPROVED** [review]
This looks good from my side.

File: clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst [hunk 11]
```diff
@@ -45,8 +47,7 @@ The following functions are always checked, regardless of *Annex K* availability
  - ``rewind``, suggested replacement: ``fseek``
  - ``setbuf``, suggested replacement: ``setvbuf``
 
-If `ReportMoreUnsafeFunctions
-<unsafe-functions.html#cmdoption-arg-ReportMoreUnsafeFunctions>`_ is enabled,
+If :ref:`ReportMoreUnsafeFunctions<option-ReportMoreUnsafeFunctions>` is enabled,
```
Comment (whisperity) [inline clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst:50]:
```suggestion
If :option:`ReportMoreUnsafeFunctions` is enabled,
```

File: clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst [hunk 12]
```diff
@@ -74,10 +75,50 @@ Both macros have to be defined to suggest replacement functions from *Annex K.*
 ``__STDC_WANT_LIB_EXT1__`` must be defined to ``1`` by the user **before**
 including any system headers.
 
+.. _CustomFunctions:
+
+Custom functions
+----------------
+
+The option `CustomFunctions` allows the user to define custom functions to be
```
Comment (whisperity) [inline clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst:83]:
```suggestion
The option :option:`CustomFunctions` allows the user to define custom functions to be
```

File: clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst [hunk 13]
```diff
@@ -74,10 +75,50 @@ Both macros have to be defined to suggest replacement functions from *Annex K.*
...
+.. code::
+
+   bugprone-unsafe-functions.CustomFunctions="
+     functionRegex1[, replacement1[, reason1]]; 
+     functionRegex2[, replacement2[, reason2]];
+     ...
+   "
+
+The functions are matched using POSIX extended regular expressions.
+*(Note: The regular expressions do not support negative* ``(?!)`` *matches)*
```
Comment (whisperity) [inline clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst:95]:
```suggestion
*(Note: The regular expressions do not support negative* ``(?!)`` *matches.)*
```

File: clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst [hunk 14]
```diff
@@ -74,10 +75,50 @@ Both macros have to be defined to suggest replacement functions from *Annex K.*
...
+If the regular expression contains the character `:`, it is matched against the
+qualified name (i.e. ``std::original``), otherwise the regex is matched against the unqualified name (``original``).
+If the regular expression starts with `::` (or `^::`), it is matched against the
+fully qualified name (``::std::original``).
 
 Options
 -------
 
+.. _option-ReportMoreUnsafeFunctions:
+
```
Comment (whisperity) [inline clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst:121]:
```suggestion
```

With using `:option:...` maybe these are not needed, but please double check!

File: clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst [hunk 15]
```diff
@@ -86,6 +127,21 @@ Options
    this option enables.
    Default is `true`.
 
+.. _option-ReportDefaultFunctions:
+
```
Comment (whisperity) [inline clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst:131]:
```suggestion
```

File: clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst [hunk 16]
```diff
@@ -86,6 +127,21 @@ Options
...
+.. option:: ReportDefaultFunctions
+
+    When `true`, the check reports the default set of functions.
+    Copnsider changing the setting to false if you only want to see custom
```
Comment (whisperity) [inline clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst:135]:
```suggestion
    Consider changing the setting to false if you only want to see custom
```

Typo!

**Review by whisperity — APPROVED** [review]

**Comment by Discookie** [comment]
Fixed the outstanding minor issues.
